src/coreclr/jit/lclvars.cpp

```diff
@@ -2045,8 +2045,21 @@ bool Compiler::StructPromotionHelper::ShouldPromoteStructVar(unsigned lclNum)
         // multiple registers?
         if (compiler->lvaIsMultiregStruct(varDsc, compiler->info.compIsVarArgs))
         {
-            if ((structPromotionInfo.fieldCnt != 2) &&
-                !((structPromotionInfo.fieldCnt == 1) && varTypeIsSIMD(structPromotionInfo.fields[0].fldType)))
+            if (structPromotionInfo.fieldCnt == 2)
+            {
+                unsigned structSize = varDsc->GetLayout()->GetSize();
+                if (structPromotionInfo.fields[0].fldSize * 2 != structSize)
```

Comment:
Wouldn't this be better and more general?
```suggestion
                if (structPromotionInfo.containsHoles)
```

(e.g., why `*2` instead of `fields[0].fldSize + fields[1].fldSize`?)

Seems like the next condition maybe should also check `containsHoles` in case there's an explicit layout struct of a single SIMD field where the explicit layout is larger than the SIMD field

Reply:
Ah, I actually had `fields[0].fldSize + fields[1].fldSize` but somehow I didn't push that change. Let me do it.

Reply:
Looking at the code around, `containsHoles` does exactly that, so I should just replace this check with `containHoles`. Perhaps I should also add a check for `customLayout` the way it is done here:

https://github.com/dotnet/runtime/blob/7dcb2892c11d4ab66ba72b767783231e699b28f3/src/coreclr/jit/lclvars.cpp#L2018-L2021

Reply:
Yes, probably should add `&& structPromotionInfo.customLayout` as well. I think the logic here is: a default/automatic struct layout might end up with holes due to alignment padding, but users can't depend on values in that padding. However, if the user explicitly set a custom/explicit layout, then we can't depend on the padding/holes being uninteresting.